### PR TITLE
[p2p/stream] Move `p2p/authenticated/connection` to new `stream` crate

### PR DESCRIPTION
**Pull request by BrendanChou**
fixes https://github.com/commonwarexyz/monorepo/issues/139

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -17,6 +17,7 @@ commonware-macros = { version = "0.0.4", path = "macros" }
 commonware-p2p = { version = "0.0.23", path = "p2p" }
 commonware-runtime = { version = "0.0.9", path = "runtime" }
 commonware-storage = { version = "0.0.4", path = "storage" }
+commonware-stream = { version = "0.0.1", path = "stream" }
```
Comment (patrick-ogrady) [inline Cargo.toml:20]: use `0.0.2` here (`0.0.1` has already been pushed)

File: stream/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,33 @@
+[package]
+name = "commonware-stream"
+edition = "2021"
+publish = true
+version = "0.0.1"
```
Comment (patrick-ogrady) [inline stream/Cargo.toml:5]: Use `0.0.2`

File: stream/src/connection/wire.proto [hunk 3]
```diff
@@ -0,0 +1,33 @@
+syntax = "proto3";
+package wire;
+
+// Message is the only allowed message format that can be sent between peers.
```
Comment (patrick-ogrady) [inline stream/src/connection/wire.proto]: nit: let's remove this nesting given there is only 1 message type

File: stream/src/lib.rs
Comment (patrick-ogrady) [inline stream/src/lib.rs:1]: I believe most (if not all) should stay in `p2p`

File: stream/src/connection/mod.rs
Comment (patrick-ogrady) [inline stream/src/connection/mod.rs:1]: Now that stream is standalone, we'll probably want to copy/move over everything I wrote about how handshake/stream encryption works.

Comment (patrick-ogrady) [inline stream/src/connection/mod.rs:1]: Pondering what the best name for this `dialect` would be....(`connection` is a placeholder)

File: p2p/src/authenticated/actors/peer/actor.rs [hunk 4]
```diff
@@ -102,14 +103,14 @@ impl<E: Spawner + Clock + ReasonablyRealtime + Rng + CryptoRng> Actor<E> {
                         content,
                     }
                 })),
-            };
+            }.encode_to_vec();
```
Comment (patrick-ogrady) [inline p2p/src/authenticated/actors/peer/actor.rs]: You can do `.encode_to_vec().into()` and remove the `Bytes::from(..)` part below.

File: p2p/src/authenticated/actors/dialer.rs [hunk 5]
```diff
@@ -24,7 +27,7 @@ use tracing::debug;
 
 pub struct Config<C: Scheme> {
     pub registry: Arc<Mutex<Registry>>,
-    pub connection: connection::Config<C>,
+    pub connection: ConnectionConfig<C>,
```
Comment (patrick-ogrady) [inline p2p/src/authenticated/actors/dialer.rs:30]: Agree that this is probably a good switch (from using `connection::Config` and importing `connection::self`)

File: p2p/src/authenticated/actors/peer/actor.rs [hunk 6]
```diff
@@ -102,14 +103,14 @@ impl<E: Spawner + Clock + ReasonablyRealtime + Rng + CryptoRng> Actor<E> {
...
             trace!(
                 part = part,
                 total_parts = total_parts - 1,
                 peer = hex(peer),
                 "sending chunk",
             );
-            sender.send(msg).await.map_err(Error::SendFailed)?;
+            sender.send(Bytes::from(msg)).await.map_err(Error::SendFailed)?;
```
Comment (patrick-ogrady) [inline p2p/src/authenticated/actors/peer/actor.rs]: You shouldn't need this wrapper. Use `.encode_to_vec().into()` above (like you do below).

File: p2p/src/authenticated/actors/peer/mod.rs [hunk 7]
```diff
@@ -33,8 +35,8 @@ pub enum Error {
     PeerDisconnected,
     #[error("receive failed: {0}")]
     ReceiveFailed(connection::Error),
-    #[error("unexpected handshake message")]
-    UnexpectedHandshake,
+    #[error("decode failed: {0}")]
+    DecodeFailed(DecodeError),
```
Comment (patrick-ogrady) [inline p2p/src/authenticated/actors/peer/mod.rs:39]: Nice...I was just masking all decoding errors before.

File: p2p/src/authenticated/actors/peer/actor.rs [hunk 8]
```diff
@@ -355,12 +364,6 @@ impl<E: Spawner + Clock + ReasonablyRealtime + Rng + CryptoRng> Actor<E> {
                                 debug!(err=?e, "failed to send message to client");
                             }
                         }
-                        Some(wire::message::Payload::Handshake(_)) => {
-                            self.received_messages
```
Comment (patrick-ogrady) [inline p2p/src/authenticated/actors/peer/actor.rs:359]: You'll also need to remove some now unnecessary items in `metrics.rs`: https://github.com/commonwarexyz/monorepo/blob/fcafa8376d5c1fdcfe99ee56fb828a11e5a42dec/p2p/src/authenticated/metrics.rs#L24

File: stream/src/connection/utils.rs [hunk 9]
```diff
@@ -5,7 +5,9 @@ pub fn nonce_bytes(dialer: bool, iter: u16, seq: u64) -> Nonce {
     if dialer {
         nonce_bytes[0] = 0b10000000; // Set the first bit of the byte
     }
-    nonce_bytes[2..4].copy_from_slice(&iter.to_be_bytes());
+    if iter > 0 {
```
Comment (patrick-ogrady) [inline stream/src/connection/utils.rs]: Nice low-hanging fruit.

File: stream/src/connection/instance.rs [hunk 10]
```diff
@@ -12,7 +11,6 @@ use chacha20poly1305::{
 use commonware_cryptography::{PublicKey, Scheme};
 use commonware_macros::select;
 use commonware_runtime::{Clock, Sink, Spawner, Stream};
-use prost::Message;
 use rand::{CryptoRng, Rng};
 
 const CHUNK_PADDING: usize = 64 /* protobuf overhead */ + 12 /* chunk info */ + 16 /* encryption tag */;
```
Comment (patrick-ogrady) [inline stream/src/connection/instance.rs]: chunks are something implemented by `p2p` (not stream), we should think about how to best handle this const.

File: stream/src/connection/mod.rs [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+mod wire {
+    include!(concat!(env!("OUT_DIR"), "/wire.rs"));
+}
+mod x25519;
+
+pub use handshake::IncomingHandshake;
+pub use instance::{Instance, Sender};
+
+#[derive(Clone)]
+pub struct Config<C: Scheme> {
```
Comment (patrick-ogrady) [inline stream/src/connection/mod.rs]: Will probably also want to accept metrics here?

File: stream/src/connection/mod.rs [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+    pub crypto: C,
+    pub namespace: Vec<u8>,
+    pub max_message_size: usize,
+    pub synchrony_bound: Duration,
```
Comment (patrick-ogrady) [inline stream/src/connection/mod.rs]: Should put comments for these configs now that they are public.

File: stream/src/connection/instance.rs (hunk 10 above)
Comment (BrendanChou) [reply stream/src/connection/instance.rs]: Fixed by https://github.com/commonwarexyz/monorepo/issues/185

File: stream/src/connection/mod.rs (hunk 11 above)
Comment (BrendanChou) [reply stream/src/connection/mod.rs]: Fixed by https://github.com/commonwarexyz/monorepo/issues/184

File: stream/src/connection/mod.rs
Comment (BrendanChou) [reply stream/src/connection/mod.rs:1]: Changed to `placeholder` [here](https://github.com/commonwarexyz/monorepo/pull/182/commits/54ac7c6c999f950b313742ba8a4add1237782a7c)

File: stream/README.md [hunk 13]
```diff
@@ -0,0 +1,10 @@
+# commonware-stream
+
+[![Crates.io](https://img.shields.io/crates/v/commonware-stream.svg)](https://crates.io/crates/commonware-stream)
+[![Docs.rs](https://docs.rs/commonware-stream/badge.svg)](https://docs.rs/commonware-stream)
+
+Connect to peers using cryptographically-authenticated connections.
```
Comment (patrick-ogrady) [inline stream/README.md]: nit: for now just put placeholder/tbd here (and the places where this string exists)...I think we'll want to put a bit more time into this description.

File: stream/Cargo.toml [hunk 14]
```diff
@@ -0,0 +1,27 @@
+[package]
+name = "commonware-stream"
+edition = "2021"
+publish = true
+version = "0.0.2"
```
Comment (patrick-ogrady) [inline stream/Cargo.toml]: Change this back to 0.0.1 to avoid triggering a release

File: Cargo.toml [hunk 15]
```diff
@@ -17,6 +17,7 @@ commonware-macros = { version = "0.0.4", path = "macros" }
 commonware-p2p = { version = "0.0.23", path = "p2p" }
 commonware-runtime = { version = "0.0.9", path = "runtime" }
 commonware-storage = { version = "0.0.4", path = "storage" }
+commonware-stream = { version = "0.0.2", path = "stream" }
```
Comment (patrick-ogrady) [inline Cargo.toml]: You'll have to change to to 0.0.1 too

**Comment by patrick-ogrady** [comment]
Can you leave a TODO above that `CHUNK_PADDING` const (with a link to the issue)?

File: stream/src/lib.rs [hunk 16]
```diff
@@ -0,0 +1,8 @@
+//! Communicate with authenticated peers over encrypted connections.
```
Comment (patrick-ogrady) [inline stream/src/lib.rs]: nit: placeholder description.

File: stream/src/placeholder/mod.rs [hunk 17]
```diff
@@ -0,0 +1,165 @@
...
+mod utils;
+mod wire {
+    include!(concat!(env!("OUT_DIR"), "/wire.rs"));
+}
+mod x25519;
+
+pub use handshake::IncomingHandshake;
+pub use instance::{Instance, Sender};
+
+
```
Comment (patrick-ogrady) [inline stream/src/placeholder/mod.rs]: nit extra `\n`

**Review by patrick-ogrady — APPROVED** [review]

**Pull request merged**